pkg/controller/cnp/clusternodeprediction_controller.go

```diff
@@ -0,0 +1,114 @@
+package cnp
+
+import (
+	"context"
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/version"
+	"k8s.io/client-go/scale"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	predictionapi "github.com/gocrane/api/prediction/v1alpha1"
+)
+
+type ClusterNodePredictionController struct {
+	client.Client
+	Log         logr.Logger
+	Scheme      *runtime.Scheme
+	RestMapper  meta.RESTMapper
+	Recorder    record.EventRecorder
+	scaleClient scale.ScalesGetter
+	K8SVersion  *version.Version
+}
+
+func (c *ClusterNodePredictionController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// Log for controller information
+	c.Log.Info("got", "cnp", req.NamespacedName)
+
+	// Get cnp object, Ignore object not found event
+	var cnp predictionapi.ClusterNodePrediction
+	if err := c.Client.Get(ctx, req.NamespacedName, &cnp); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Match nodes according to labels
+	var nodeList v1.NodeList
+	var matchingLabels client.MatchingLabels = cnp.Spec.NodeSelector
+	opts := []client.ListOption{
+		matchingLabels,
+	}
+	if err := c.Client.List(ctx, &nodeList, opts...); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// When labels no match any node
+	// Set the status, Then reconcile func end
+	if len(nodeList.Items) == 0 {
+		status := predictionapi.ClusterNodePredictionStatus{
+			CurrentNumberCreated: 0,
+			DesiredNumberCreated: 0,
+			Conditions: nil,
+		}
+		if err := c.updateStatus(ctx, &cnp, status); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Mutate TimeSeriesPrediction according to the number of node list
+	var successCount int
+	for _, node := range nodeList.Items {
```

Comment:
seems when never there is an error, the entire sync loop exit without updating the status? it's better if we can support partial success status, as the actual tsp is created already.